### PR TITLE
feat: event and workflow run metadata

### PR DESCRIPTION
**Pull request by grutt**
# Description

Often arbitrary metadata tags or labels are desirable to organize events and workflow runs. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## What's Changed

- [x] Adds additionalMetadata columns to event and workflow runs
- [x] Revise list queries to filter by key-value pairs
- [x] add supporting ui
- [x] adds metadata to go sdk event push 

References: https://github.com/hatchet-dev/hatchet-typescript/pull/86

### CODE REVIEW

File: internal/repository/event.go [hunk 1]
```diff
@@ -48,6 +51,9 @@ type ListEventOpts struct {
 
 	// (optional) the order direction
 	OrderDirection *string `validate:"omitempty,oneof=ASC DESC"`
+
+	// (optional) the event metadata
+	AdditionalMetadata *[]byte
```
Comment (abelanger5) [inline internal/repository/event.go]:
```suggestion
	AdditionalMetadata []byte
```

Arrays are nil by default, so `*` is unnecessary

File: internal/repository/workflow_run.go [hunk 2]
```diff
@@ -231,6 +253,9 @@ type ListWorkflowRunsOpts struct {
 
 	// (optional) the order direction
 	OrderDirection *string `validate:"omitempty,oneof=ASC DESC"`
+
+	// (optional) exact metadata to filter by
+	AdditionalMetadata *map[string]interface{} `validate:"omitempty"`
```
Comment (abelanger5) [inline internal/repository/workflow_run.go]:
```suggestion
	AdditionalMetadata map[string]interface{} `validate:"omitempty"`
```

File: internal/repository/workflow_run.go [hunk 3]
```diff
@@ -248,6 +273,9 @@ type WorkflowRunsMetricsOpts struct {
 
 	// (optional) the event id that triggered the workflow run
 	EventId *string `validate:"omitempty,uuid"`
+
+	// (optional) exact metadata to filter by
+	AdditionalMetadata *map[string]interface{} `validate:"omitempty"`
```
Comment (abelanger5) [inline internal/repository/workflow_run.go]:
```suggestion
	AdditionalMetadata map[string]interface{} `validate:"omitempty"`
```

File: internal/services/controllers/events/controller.go [hunk 4]
```diff
@@ -146,10 +147,18 @@ func (ec *EventsControllerImpl) handleTask(ctx context.Context, task *msgqueue.M
 		return fmt.Errorf("could not decode task metadata: %w", err)
 	}
 
-	return ec.processEvent(ctx, metadata.TenantId, payload.EventId, payload.EventKey, []byte(payload.EventData))
+	additionalMetadata := make(map[string]interface{})
+
+	err = json.Unmarshal([]byte(payload.EventAdditionalMetadata), &additionalMetadata)
```
Comment (abelanger5) [inline internal/services/controllers/events/controller.go]: Need nil check

File: prisma/migrations/20240426175227_v0_22_1/migration.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/*
```
Comment (abelanger5) [inline prisma/migrations/20240426175227_v0_22_1/migration.sql]: Collapse into 1 migration

File: api/v1/server/handlers/workflows/trigger.go [hunk 6]
```diff
@@ -17,6 +17,8 @@ import (
 )
 
 func (t *WorkflowService) WorkflowRunCreate(ctx echo.Context, request gen.WorkflowRunCreateRequestObject) (gen.WorkflowRunCreateResponseObject, error) {
+	// TODO metadata
```
Comment (abelanger5) [inline api/v1/server/handlers/workflows/trigger.go]: Remove todo?

**Review by abelanger5 — CHANGES_REQUESTED** [review]

File: api/v1/server/handlers/workflows/trigger.go [hunk 7]
```diff
@@ -57,8 +59,26 @@ func (t *WorkflowService) WorkflowRunCreate(ctx echo.Context, request gen.Workfl
 		), nil
 	}
 
-	createOpts, err := repository.GetCreateWorkflowRunOptsFromManual(workflowVersion, inputBytes)
+	additionalMetadata := make(map[string]interface{})
```
Comment (steebchen) [inline api/v1/server/handlers/workflows/trigger.go]:
```suggestion
	var additionalMetadata map[string]interface{}
```

File: api/v1/server/oas/transformers/workflow_run.go [hunk 8]
```diff
@@ -296,16 +296,28 @@ func ToWorkflowRunFromSQLC(row *dbsqlc.ListWorkflowRunsRow) *gen.WorkflowRun {
 
 	workflowRunId := sqlchelpers.UUIDToStr(run.ID)
 
+	var additionalMetadata (map[string]interface{})
```
Comment (steebchen) [inline api/v1/server/oas/transformers/workflow_run.go]:
```suggestion
	var additionalMetadata map[string]interface{}
```

File: prisma/schema.prisma [hunk 9]
```diff
@@ -226,6 +226,9 @@ model Event {
   // data stored in the event
   data Json?
 
+  // metadata stored in the event
+  additionalMetadata Json?
```
Comment (steebchen) [inline prisma/schema.prisma:233]: can we not name this `metadata`? Would be a little cleaner. Or is this conflicting with existing metadata names?

Comment (abelanger5) [reply prisma/schema.prisma:233]: I'm going to leave this as-is, we can certainly change the column name later. I imagine this is preventing `metadata` from being an overloaded field in the API. I think a more apt name would be `userMetadata`?

**Review by abelanger5 — APPROVED** [review]

**Pull request merged**